File: src/test/kotlin/tracking/steps/PreprocessorTest.kt
package tracking.steps

import org.junit.jupiter.api.DynamicTest.dynamicTest
import org.junit.jupiter.api.TestFactory
import kotlin.test.Test
import kotlin.test.assertEquals

class PreprocessorTest {
    private val subject = Preprocessor()

    @TestFactory
    fun `Cleaner regexes are working`() = mapOf(
        "***open***" to "open",
        "!closed!" to "closed",
        " ❗&nbsp;" to " ! ",
    ).map { (input, expected) ->
        dynamicTest("Test input: '${input}'") {
            val result = subject.preprocess(input, listOf())

            assertEquals(expected, result)
        }
    }

    @Test
    fun `Input gets converted to lowercase`() {
        val result = subject.preprocess("AaBbCcDdEeFf", listOf())

        assertEquals("aabbccddeeff", result)
    }

    @TestFactory
    fun `Creator aliases are getting replaced with the name placeholder`() = listOf(
        Triple(
            "An Intergalactic House of Pancakes work",
            listOf("Intergalactic House of Pancakes"),
            "an CREATOR_NAME work",
        ),
        Triple(
            "An Intergalactic House of Pancake's work",
            listOf("Intergalactic House of Pancakes"),
            "an CREATOR_NAME work",
        ),
        Triple(
            "About Intergalactic Pancake's work",
            listOf("Intergalactic Pancake"),
            "about CREATOR_NAME's work",
        ),
        Triple( // Multiple aliases, 's form, case-insensitive, "creator" in aliases
            "asdf Studio's uiop Creator asdf Studios zxcv",
            listOf("StUdIoS", "cReatOR"),
            "asdf CREATOR_NAME uiop CREATOR_NAME asdf CREATOR_NAME zxcv",
        ),
    ).map { (input, aliases, expected) ->
        dynamicTest("Test input: '${input}'") {
            val result = subject.preprocess(input, aliases)

            assertEquals(expected, result)
        }
    }

    @TestFactory
    fun `False positives are being removed`() = mapOf(
        "even though you're closed for commissions" to "",
        "while mine commissions are open" to "",
        "if my quotes open" to "",
        "- art commissions are open" to "- are open",
        "after the commissions close" to "",
        "although comms are closed" to "",
        "as soon as we're open" to "",
        "next commissions opening" to "",
        "commissions: open January" to "",
        "The Creator is now opening for quotes a few weeks before commission slots open" to "",
        "when do you open for" to "",
        "when i'm taking" to "",
        "when will you start taking new commissions?" to "?",
    ).map { (input, expected) ->
        dynamicTest("Test input: '${input}'") {
            val result = subject.preprocess(input, listOf("The Creator"))

            assertEquals(expected, result)
        }
    }
}

File: symfony/tests/Tracking/TextProcessing/PreprocessorTest.php
<?php

declare(strict_types=1);

namespace App\Tests\Tracking\TextProcessing;

use App\Tests\TestUtils\Cases\FuzzrakeTestCase;
use App\Tests\TestUtils\DataDefinitions;
use App\Tracking\Patterns\Patterns;
use App\Tracking\Patterns\RegexesLoader;
use App\Tracking\TextProcessing\Preprocessor;
use Override;
use PHPUnit\Framework\Attributes\DataProvider;
use PHPUnit\Framework\Attributes\Small;
use Psr\Log\LoggerInterface;
use Veelkoov\Debris\StringList;

#[Small]
class PreprocessorTest extends FuzzrakeTestCase
{
    private Preprocessor $subject;

    #[Override]
    public function setUp(): void
    {
        /* @phpstan-ignore argument.type */
        $regexesLoader = new RegexesLoader(DataDefinitions::get('tracking.yaml', 'tracking'));
        $this->subject = new Preprocessor(self::createStub(LoggerInterface::class), new Patterns($regexesLoader));
    }

    /**
     * @return list<array{string, string}>
     */
    public static function cleanerRegexesAreWorkingDataProvider(): array
    {
        return [
            ['***open***', 'open'],
            ['!closed!', 'closed'],
            [' ❗&nbsp;', ' ! '], // Unicode NBSP, emoticon !, HTML entity NBSP
            ["\t", ' '],
        ];
    }

    #[DataProvider('cleanerRegexesAreWorkingDataProvider')]
    public function testCleanerRegexesAreWorking(string $input, string $expected): void
    {
        $result = $this->subject->getPreprocessedContent(self::getAnalysisInput(contents: $input));

        self::assertSame($expected, $result);
    }

    public function testInputGetsConvertedToLowercase(): void
    {
        $result = $this->subject->getPreprocessedContent(self::getAnalysisInput(contents: 'AaBbCcDdEeFf'));

        self::assertSame('aabbccddeeff', $result);
    }

    /**
     * @return list<array{string, StringList, string}>
     */
    public static function creatorAliasesAreGettingReplacedWithTheNamePlaceholderDataProvider(): array
    {
        return [
            [
                'An Intergalactic House of Pancakes work',
                new StringList(['Intergalactic House of Pancakes']),
                'an CREATOR_NAME work',
            ],
            [
                "An Intergalactic House of Pancake's work",
                new StringList(['Intergalactic House of Pancakes']),
                'an CREATOR_NAME work',
            ],
            [
                "About Intergalactic Pancake's work",
                new StringList(['Intergalactic Pancake']),
                "about CREATOR_NAME's work",
            ],
            [
                // Multiple aliases, 's form, case-insensitive, "creator" in aliases
                "asdf Studio's uiop Creator asdf Studios zxcv",
                new StringList(['StUdIoS', 'cReatOR']),
                'asdf CREATOR_NAME uiop CREATOR_NAME asdf CREATOR_NAME zxcv',
            ],
        ];
    }

    #[DataProvider('creatorAliasesAreGettingReplacedWithTheNamePlaceholderDataProvider')]
    public function testAliasesGetReplacedWithPlaceholder(string $input, StringList $aliases, string $expected): void
    {
        $result = $this->subject->getPreprocessedContent(self::getAnalysisInput(name: $aliases->at(0),
            formerly: $aliases->slice(1)->getValuesArray(), contents: $input));

        self::assertSame($expected, $result);
    }

    /**
     * @return list<array{string, string}>
     */
    public static function falsePositivesAreBeingRemovedDataProvider(): array
    {
        return [
            ["even though you're closed for commissions", 'FALSE_POSITIVE'],
            ['while mine commissions are open', 'FALSE_POSITIVE'],
            ['if my quotes open', 'FALSE_POSITIVE'],
            ['- art commissions are open', '- FALSE_POSITIVE are open'],
            ['after the commissions close', 'FALSE_POSITIVE'],
            ['although comms are closed', 'FALSE_POSITIVE'],
            ["as soon as we're open", 'FALSE_POSITIVE'],
            ['next commissions opening', 'FALSE_POSITIVE'],
            ['commissions: open January', 'FALSE_POSITIVE'],
            ['The Creator is now opening for quotes a few weeks before commission slots open', 'FALSE_POSITIVE'],
            ['when do you open for', 'FALSE_POSITIVE'],
            ["when i'm taking", 'FALSE_POSITIVE'],
            ['when will you start taking new commissions?', 'FALSE_POSITIVE?'],
        ];
    }

    #[DataProvider('falsePositivesAreBeingRemovedDataProvider')]
    public function testFalsePositivesAreBeingRemoved(string $input, string $expected): void
    {
        $result = $this->subject->getPreprocessedContent(self::getAnalysisInput(contents: $input, name: 'The Creator'));

        self::assertSame($expected, $result);
    }

    public function testLongContentsGetsTruncated(): void
    {
        $input = 'hello'.str_repeat(str_repeat('1234567890abcdef', 64), 1024).'bye';

        $result = $this->subject->getPreprocessedContent(self::getAnalysisInput(contents: $input));

        self::assertSame(1048576, mb_strlen($result));
        self::assertStringStartsWith('hello', $result);
        self::assertStringEndsNotWith('bye', $result);
    }
}
